feat(cli): add persistent peers

MegaRedHand opened the pull request:
Closes: https://github.com/paradigmxyz/reth/issues/745

This PR introduces:

- kill signal handling via _tokio_'s `ctrl_c`
- peer list dump on node termination
- peer loading on startup

Related to: https://github.com/paradigmxyz/reth/issues/807 https://github.com/paradigmxyz/reth/issues/1036

gakonst commented on bin/reth/src/node/mod.rs:
```diff
@@ -259,6 +272,32 @@ impl Command {
+            Err(e) if e.kind() == ErrorKind::NotFound => return Ok(()),
+            Err(e) => Err(e)?,
+        };
+        info!(target: "reth::cli", file = ?self.known, "Loading saved peers");
+        let known_peers: Vec<NodeRecord> = serde_json::from_reader(reader)?;
+        known_peers.into_iter().for_each(|record| network.add_peer(record.id, record.udp_addr()));
+        Ok(())
+    }
+
+    // Dumps peers to `file_path` for persistence.
+    async fn dump_peers(&self, network: &NetworkHandle) -> Result<(), eyre::Error> {
+        info!(target: "reth::cli", file = ?self.known, "Saving current peers");
+        let writer = std::io::BufWriter::new(std::fs::File::create(&self.known)?);
+        let known_peers: Vec<NodeRecord> = network
+            .get_peers()
+            .await?
+            .into_iter()
+            .map(|PeerInfo { remote_id, remote_addr, .. }| NodeRecord::new(remote_addr, remote_id))
+            .collect();
+        serde_json::to_writer_pretty(writer, &known_peers)?;
```
Can we make this 
1. opt-out
2. store the reputation of the peers?
3. only save peers above x% reputation based on the max?

mattsse replied:
> store the reputation of the peers?

I don't think this is necessary, because the peers here don't include peers that we know are useless (different network for example) and reputation is merely used to rank peers or put them in timeout if they are busy at the moment, which is perhaps no longer the case when the node is restarting.

MegaRedHand replied:
Added a cli flag for disabling peer persistence. Also added peer filtering based on reputation. For now it saves only peers with reputation higher than max peer reputation minus `10000`

gakonst reviewed (changes requested):
nice! thank you for taking this on.

mattsse commented on bin/reth/src/node/mod.rs:
```diff
@@ -122,7 +130,12 @@ impl Command {
         // Run pipeline
         info!(target: "reth::cli", "Starting sync pipeline");
-        pipeline.run(db.clone()).await?;
+        tokio::select! {
+            res = pipeline.run(db.clone()) => res?,
+            _ = tokio::signal::ctrl_c() => {},
+        };
```
for now this looks fine since the pipline is run on the main task.

but ideally, this should be integrated via #1036

mattsse commented on bin/reth/src/node/mod.rs:
```diff
@@ -259,6 +272,32 @@ impl Command {
         )
     }
+
+    async fn load_peers(&self, network: &NetworkHandle) -> Result<(), eyre::Error> {
+        let reader = match std::fs::File::open(&self.known) {
+            Ok(file) => std::io::BufReader::new(file),
+            Err(e) if e.kind() == ErrorKind::NotFound => return Ok(()),
+            Err(e) => Err(e)?,
+        };
+        info!(target: "reth::cli", file = ?self.known, "Loading saved peers");
+        let known_peers: Vec<NodeRecord> = serde_json::from_reader(reader)?;
+        known_peers.into_iter().for_each(|record| network.add_peer(record.id, record.udp_addr()));
+        Ok(())
+    }
```
I want this to be a standalone function that takes the path as an argument.

also please add docs.

MegaRedHand replied:
Made both standalone, and added missing docs.

mattsse commented on bin/reth/src/node/mod.rs:
```diff
@@ -259,6 +272,32 @@ impl Command {
         )
     }
+
+    async fn load_peers(&self, network: &NetworkHandle) -> Result<(), eyre::Error> {
```
```suggestion
    async fn load_peers(&self, network: &NetworkHandle) -> eyre::Result<()> {
```

mattsse commented on bin/reth/src/node/mod.rs:
```diff
@@ -259,6 +272,32 @@ impl Command {
+    }
+
+    // Dumps peers to `file_path` for persistence.
+    async fn dump_peers(&self, network: &NetworkHandle) -> Result<(), eyre::Error> {
```
same

mattsse commented on bin/reth/src/node/mod.rs:
```diff
@@ -46,6 +46,11 @@ pub struct Command {
 
+    /// The path to the known peers file. Connected peers are
+    /// dumped to this file on node shutdown, and read on startup.
+    #[arg(long, value_name = "FILE", verbatim_doc_comment, default_value_t)]
+    known: PlatformPath<KnownPeersPath>,
```
`--known` as argument is a bit weird imo

the name should make it obvious that this is using peers written to disk.

MegaRedHand replied:
Maybe `--peers-file`?

mattsse replied:
that's a good suggestion.

hmm, I guess it would make sense to provide an option for regular peers as well in `NetworkConfig` and then a helper function to read from disk?

And yes, these functions definitely fit better in the network crate.

MegaRedHand replied:
Renamed it to `--peers-file` and moved the cli flags to `NetworkOpts` in `bin/reth/src/lib.rs`. Also added a function to load peers from file in `NetworkConfig`.

gakonst commented on bin/reth/src/node/mod.rs:
```diff
@@ -115,6 +115,11 @@ impl Command {
         let netconf = self.load_network_config(&config, &db);
         let network = netconf.start_network().await?;
+
+        if !self.network.no_persist_peers {
+            load_peers(&self.network.peers_file, &network).await?;
+        }
```
I would probably pull this in the `start_network` fn or make it a method on `NetworkHandle`

MegaRedHand replied:
Added this to `NetworkConfig`, and `load_network_config`.

gakonst commented on bin/reth/src/node/mod.rs:
```diff
@@ -271,6 +283,32 @@ impl Command {
+    Ok(())
+}
+
+/// Dumps peers to `file_path` for persistence.
+async fn dump_peers(file_path: &impl AsRef<Path>, network: &NetworkHandle) -> eyre::Result<()> {
+    info!(target: "reth::cli", file = %file_path.as_ref().display(), "Saving current peers");
+    let writer = std::io::BufWriter::new(std::fs::File::create(file_path)?);
+    let known_peers = network.peers_handle().all_peers().await;
+
+    let threshold = known_peers.iter().map(|(rep, _)| *rep).max().unwrap_or(0) - 10000;
+    let top_peers: Vec<NodeRecord> =
+        known_peers.into_iter().filter(|(rep, _)| *rep > threshold).map(|(_, node)| node).collect();
+    serde_json::to_writer_pretty(writer, &top_peers)?;
+    Ok(())
+}
```
Nit: could move this (along with the if check) to the Drop impl of `NetworkHandle`, this would make sure that it's always used

gakonst commented on crates/net/network/src/peers/manager.rs:
```diff
@@ -63,6 +63,14 @@ impl PeersHandle {
         rx.await.unwrap_or(None)
     }
+
+    /// Send a reputation change for the given peer.
```
```suggestion
    /// Returns all the currently connected peers and their reputations.
```

gakonst commented on bin/reth/src/node/mod.rs:
```diff
@@ -271,6 +283,32 @@ impl Command {
+    let writer = std::io::BufWriter::new(std::fs::File::create(file_path)?);
+    let known_peers = network.peers_handle().all_peers().await;
+
+    let threshold = known_peers.iter().map(|(rep, _)| *rep).max().unwrap_or(0) - 10000;
```
replace the `10_000` with any of the reputation cosntatns, or add note on why this is chosen

MegaRedHand replied:
Just changed it to something closer to your initial suggestion. wdyt?

MegaRedHand commented:
Ended up moving the `dump_peers` function into `PeersManager`, and added a `Drop` implementation that optionally persists peers. Also moved the `load_peers` function to `PeersConfig`.

gakonst reviewed (approved):
good from me - @mattsse ?

mattsse commented on bin/reth/src/node/mod.rs:
```diff
@@ -181,12 +184,20 @@ impl Command {
+            None
+        } else {
+            let file_path = &self.network.peers_file;
+            info!(target: "reth::cli", file = %file_path.as_ref().display(), "Loading saved peers");
```
this is not accurate because it does not actually load here.

mattsse commented on crates/net/network/src/peers/manager.rs:
```diff
@@ -992,6 +1064,35 @@ impl PeersConfig {
         self
     }
+
+    /// Nodes to initially connect to.
```
please rephrase, these are nodes that are available at launch

mattsse commented on crates/net/network/src/peers/manager.rs:
```diff
@@ -992,6 +1064,35 @@ impl PeersConfig {
+    }
+
+    /// Adds file to save peers to on shutdown
+    pub fn with_optional_peers_file(mut self, peers_file: Option<impl AsRef<Path>>) -> Self {
```
```suggestion
    pub fn with_peers_file(mut self, peers_file: Option<Into<PathBuf>>) -> Self {
```

mattsse commented on crates/net/network/src/peers/manager.rs:
```diff
@@ -648,6 +676,31 @@ impl PeersManager {
+    fn dump_peers(&mut self, file_path: &PathBuf) -> Result<(), io::Error> {
+        info!(target : "net::peers", file = %file_path.display(), "Saving current peers");
+        let writer = std::io::BufWriter::new(std::fs::File::create(file_path)?);
+        let known_peers: Vec<(i32, NodeRecord)> =
+            self.peers.iter().map(|(k, v)| (v.reputation, NodeRecord::new(v.addr, *k))).collect();
+
+        let discard_percentage = 0.7;
+
+        // Get top peer reputation or 0 if positive
+        let max_rep = known_peers.iter().map(|(rep, _)| *rep).max().unwrap_or(0).min(0);
+
+        // Discards peers below X% reputation, relative to `max_peer`
+        let threshold =
+            ((max_rep - BANNED_REPUTATION) as f64 * discard_percentage) as i32 + BANNED_REPUTATION;
+
+        let top_peers: Vec<NodeRecord> = known_peers
+            .into_iter()
+            .filter(|(rep, _)| *rep > threshold)
+            .map(|(_, node)| node)
+            .collect();
```
imo no need for this at all, can use all nodes.

mattsse commented on crates/net/network/src/peers/manager.rs:
```diff
@@ -876,6 +938,8 @@ pub(crate) enum PeerCommand {
     ReputationChange(PeerId, ReputationChangeKind),
     /// Get information about a peer
     GetPeer(PeerId, oneshot::Sender<Option<Peer>>),
+    /// Get reputation and address information on all peers
+    GetPeers(oneshot::Sender<Vec<(i32, NodeRecord)>>),
```
I'd like this to return `Vec<Peer>` instead, can be included in the `PeersInfo` trait.

MegaRedHand replied:
I need some way to retrieve the `PeerId` for each one. As it isn't in the `Peer`, I'd need this to return a `Vec<(PeerId, Peer)>`

mattsse replied:
I see, then let's do `Vec<NodeRecord>`

mattsse commented on crates/net/network/src/peers/manager.rs:
```diff
@@ -656,6 +709,15 @@ impl Default for PeersManager {
 }
 
+impl Drop for PeersManager {
+    fn drop(&mut self) {
+        let Some(file_path) = self.peers_file.take() else { return };
+        if let Err(err) = self.dump_peers(&file_path) {
+            warn!(target = "net::peers", ?err, file = ?file_path.display(), "Failed to save peers to file")
+        }
+    }
```
this won't work because it won't fire on SIGNT (ctrl-c)

gakonst replied:
crazy

MegaRedHand replied:
Really crazy 😵‍💫

mattsse commented on crates/net/network/src/peers/manager.rs:
```diff
@@ -921,6 +985,12 @@ pub struct PeersConfig {
+    pub basic_nodes: HashSet<NodeRecord>,
+    /// Optional file to persist peers in.
+    #[cfg_attr(feature = "serde", serde(skip))]
+    pub peers_file: Option<PathBuf>,
```
let's remove this option because not useful

mattsse commented on bin/reth/src/node/mod.rs:
```diff
@@ -126,7 +126,10 @@ impl Command {
-        pipeline.run(db.clone()).await?;
+        tokio::select! {
+            res = pipeline.run(db.clone()) => res?,
+            _ = tokio::signal::ctrl_c() => {},
```
this is where we need to handle things like flush to disk, so here we need to `GetPeers` write to file.

gakonst replied:
ACK - this was my bad then @MegaRedHand

MegaRedHand replied:
😄 No problem! I also didn't know about `drop` not triggering 🤯 
Just finished addressing all comments, and moved the flush back here.

mattsse reviewed (changes requested):
There's a problem with flushing on drop, because it's not guaranteed to fire.

you can easily test this by running this and terminate via `ctrl-c`

```rust
struct A;

impl Drop for A {
    fn drop(&mut self) {
        print!("A dropped")
    }
}
fn main() {
   let a = A {};
   loop {
   }
}
```

this needs to happen in the `ctrl-c` listener instead.

also, we can simplify some parts.
left some additional nits and suggestions.

mattsse commented on bin/reth/src/lib.rs:
```diff
@@ -42,4 +43,13 @@ struct NetworkOpts {
+    peers_file: PlatformPath<KnownPeersPath>,
+
+    /// Do not persist peers. This ignores the --peers-file option.
+    #[arg(long, verbatim_doc_comment)]
```
should we use `conflicts_with = "peers_file"`?

MegaRedHand replied:
That's a good idea! If both options are used, it's probably a mistake and should be notified to the user.

mattsse commented on bin/reth/src/node/mod.rs:
```diff
@@ -285,6 +294,16 @@ impl Command {
+/// Dumps peers to `file_path` for persistence.
+async fn dump_peers(file_path: &Path, network: NetworkHandle) -> Result<(), io::Error> {
+    info!(target : "net::peers", file = %file_path.display(), "Saving current peers");
+    let writer = std::io::BufWriter::new(std::fs::File::create(file_path)?);
+    let known_peers = network.peers_handle().all_peers().await;
```
since this is time sensitive, let's use the `tokio::fs::File::create` variant so we can create the file and fetch the peers concurrently via `select`

mattsse replied:
but perhaps this is not even required and we can use `fs::write` directly?

mattsse commented on crates/net/network/src/peers/manager.rs:
```diff
@@ -63,6 +64,14 @@ impl PeersHandle {
         rx.await.unwrap_or(None)
     }
+
+    /// Returns all the currently connected peers and their reputations.
```
```suggestion
    /// Returns all peers in the peerset.
```

mattsse commented on crates/staged-sync/src/config.rs:
```diff
@@ -31,10 +31,13 @@ impl Config {
+        peers_file: Option<PathBuf>,
     ) -> NetworkConfig<ShareableDatabase<DB>> {
-        let peer_config = reth_network::PeersConfig::default()
-            .with_trusted_nodes(self.peers.trusted_nodes.clone())
```
why was `with_trusted_nodes` removed here?

MegaRedHand replied:
Those config options are already set in `self.peers`, so I replaced `with_trusted_nodes` and `with_connect_trusted_nodes_only` with a call to `clone`

mattsse replied:
ah okay, misread the this change,
this is basically `clon + change` now instead of `default + merge`

ty

mattsse reviewed (changes requested):
getting there :)

mattsse reviewed (approved):
lgtm.

thank you for following through on all the requests 🙏

mattsse commented:
@MegaRedHand could please revert the Cargo.lock changes?

unsure why they're even in here.
